ProjFS.Mac/PrjFSKext/Message_Kernel.cpp

```diff
@@ -33,7 +35,47 @@ void Message_Init(
     {
         header->pathSizeBytes = 0;
     }
-    
+
+    if (nullptr != sourcePath)
+    {
+        header->sourcePathSizeBytes = strlen(sourcePath) + 1;
+    }
+    else
+    {
+        header->sourcePathSizeBytes = 0;
+    }
+
     spec->messageHeader = header;
     spec->path = path;
+    spec->sourcePath = sourcePath;
+}
+
+uint32_t Message_EncodedSize(const Message& message)
+{
+    return sizeof(*message.messageHeader) + message.messageHeader->pathSizeBytes + message.messageHeader->sourcePathSizeBytes;
+}
+
+uint32_t Message_Encode(void* buffer, const uint32_t bufferSize, const Message& message)
+{
+    uint8_t* bufferPosition = static_cast<uint8_t*>(buffer);
+    uint32_t bufferBytesRemain = bufferSize;
+    
+    assert(bufferSize >= sizeof(*message.messageHeader));
+    memcpy(bufferPosition, message.messageHeader, sizeof(*message.messageHeader));
+    bufferPosition +=    sizeof(*message.messageHeader);
+    bufferBytesRemain -= sizeof(*message.messageHeader);
+    
+    for (unsigned i = 0; i < 2; ++i)
```

Comment:
Or `extent<decltype(...)>::value` (but I'm not sure we can use that in the kernel)

Reply:
`std::extent` is part of the C++ standard library, which is not available when building kexts. But yes, we can implement our own version of it; I so far have opted not to here because templated code in previous PRs has met stiff resistance. Any suggestions for naming the file where we'd gather such utility templates?

Reply:
I actually added this helper at one point during the vnode cache implementation and didn't end up needing it and it was removed from the PR.  

If you're interested you could pull the implementation and unit tests back in:

https://github.com/microsoft/VFSForGit/blob/846c64df880160abb6cf8d2c85a590ba1c519db1/ProjFS.Mac/PrjFSKext/ArrayUtilities.hpp

https://github.com/microsoft/VFSForGit/blob/846c64df880160abb6cf8d2c85a590ba1c519db1/ProjFS.Mac/PrjFSKextTests/ArrayUtilitiesTests.mm

I wasn't thrilled with the name `ArrayUtilities` but thought it was preferable to something even more generic (e.g. UtilityFunctions.h).  If we add more helper templates we could split them up by what they're used with.

What do you think?

Reply:
I had a `KextCppUtilities.hpp` way back on an early iteration of #500 which had some bits & pieces like this in it.